### DISCUSSION TITLE
Add a logo to object via a .TGA import and extrude

### DISCUSSION THREAD

**Discussion by thsmit** [leap71/PicoGK#17] [Q&A] [answered]
Hi, I am experimenting with the code. It is going well. Complements about the code and open-sourcing it!

I want to add a logo to a part by importing a logo.tga file and extruding it. I can import the TGA with:

`TgaIo.LoadTga("logo.tga", out Image logo);`

Now I am struggling with transforming the .tga file to a voxel field and maybe extruding the logo. 
Can you help me out?

Best, Thijs

**Comment by LinKayser** [comment]
The ShapeKernel has a surface modulation function that takes an image as an input, this is the easiest way to do it (we don't have embossing on the Voxel level directly).

https://github.com/leap71/LEAP71_ShapeKernel/blob/main/ShapeKernel/Modulations/SurfaceModulation(2D).cs

Maybe @josefinelissner can show us an example :)

Lin

**Comment by josefinelissner** [comment, marked as answer]
Hi Thijs!

Images can be used as 2D-SurfaceModulation objects and inputed into any base shape that can do SurfaceModulations.
However, there is a base shape called "BaseLogoBox" that is specifically designed for the use as a Logo plate. It's internal meshing strategy will copy the pixel resolution of your image.

I am attaching a example code that showcases how to use this base shape:
[Ex_LogoShowCase.txt](https://github.com/leap71/PicoGK/files/14020733/Ex_LogoShowCase.txt)

Here is the output  of Step 1, which is the simples case.
<img width="736" alt="Screenshot 2024-01-23 at 11 58 13 AM" src="https://github.com/leap71/PicoGK/assets/44703282/3c40e197-1a20-4aa3-b6a2-55c5980fa300">

Note that there is a mapping function which correlates pixel grayscale value (normalized between 0 and 1) to an "emboss" height. A different function (e.g. Option 2) will allow you to invert the image emboss.

<img width="736" alt="Screenshot 2024-01-23 at 12 00 23 PM" src="https://github.com/leap71/PicoGK/assets/44703282/d030d35c-b647-483f-a83d-558f67f12a80">

As a Step 2 you can add a transformation to the BaseLogoBox (just like any other base shape) that allows you to e.g. transform this plate into a cylindrical coordinate space.

<img width="736" alt="Screenshot 2024-01-23 at 12 05 58 PM" src="https://github.com/leap71/PicoGK/assets/44703282/67e60be9-0dee-462a-aa42-e3221f3cf4c7">

The trafo function can also be used to flip the image relative to your base shape coordinate system or to do some more elaborate / conformal geometries. If you toggle Option 2 in the trafo function you will also have an example of how to make the "emboss" overhang conformal for 3D-printing. Note that the embossed structure does not depart perfectly perpendicular from the cylinder, but at a printable angle.

![Emboss Example](https://github.com/leap71/PicoGK/assets/44703282/07157d92-827a-4ef3-8735-8dec44746e4e)

Hope this give you some background and initial direction.
Best,
Josefine

**Reply by josefinelissner** [reply]
Remember to update the TGA Filename in my code snippet, I just put a dummy name in! @thsmit

**Reply by thsmit** [reply]
Thank you @LinKayser and @josefinelissner for the quick reply. This works perfectly.
